File: springboot/src/main/java/io/proofmark/showad/jwt/JwtHelper.java
package io.proofmark.showad.jwt;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.Clock;
import java.time.Instant;
import java.util.Base64;
import java.util.Set;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * JWT decoder that mirrors the helpers exposed by the Laravel and Next.js
 * SDKs.
 *
 * <strong>Important:</strong> this helper does NOT verify the JWT signature.
 * Signature verification is the responsibility of the ProofMark backend.
 * Locally we only inspect claims to pre-filter expired or mismatched tokens
 * before paying for a network round-trip.
 *
 * <p>Defense-in-depth: rejects tokens whose header {@code alg} is {@code none}
 * or outside the HS256/HS384/HS512/RS256/RS384/RS512/ES256/ES384 whitelist.
 */
public class JwtHelper {

    public static final String EXPECTED_ISSUER = "showad-backend";

    public static final long DEFAULT_LEEWAY_SECONDS = 60L;

    public static final Set<String> ALLOWED_ALGORITHMS = Set.of(
        "HS256", "HS384", "HS512",
        "RS256", "RS384", "RS512",
        "ES256", "ES384"
    );

    private final ObjectMapper objectMapper;
    private final Clock clock;

    public JwtHelper() {
        this(new ObjectMapper(), Clock.systemUTC());
    }

    public JwtHelper(ObjectMapper objectMapper, Clock clock) {
        this.objectMapper = objectMapper;
        this.clock = clock;
    }

    public TokenClaims decode(String token) {
        if (token == null || token.isBlank()) {
            return null;
        }
        String[] parts = token.split("\\.");
        if (parts.length != 3) {
            return null;
        }
        try {
            byte[] headerBytes = Base64.getUrlDecoder().decode(padBase64(parts[0]));
            JsonNode header = objectMapper.readTree(new String(headerBytes, StandardCharsets.UTF_8));
            if (header == null || !header.isObject() || !header.hasNonNull("alg")) {
                return null;
            }
            String alg = header.get("alg").asText();
            if (!ALLOWED_ALGORITHMS.contains(alg)) {
                return null;
            }

            byte[] payload = Base64.getUrlDecoder().decode(padBase64(parts[1]));
            return objectMapper.readValue(new String(payload, StandardCharsets.UTF_8), TokenClaims.class);
        } catch (Exception ex) {
            return null;
        }
    }

    public boolean isExpired(String token) {
        return isExpired(token, DEFAULT_LEEWAY_SECONDS);
    }

    public boolean isExpired(String token, long leewaySeconds) {
        TokenClaims claims = decode(token);
        if (claims == null) {
            return true;
        }
        long now = clock.instant().getEpochSecond();
        if (claims.getExp() != null && (claims.getExp() + leewaySeconds) < now) {
            return true;
        }
        if (claims.getNbf() != null && (claims.getNbf() - leewaySeconds) > now) {
            return true;
        }
        if (claims.getIat() != null && (claims.getIat() - leewaySeconds) > now) {
            return true;
        }
        return false;
    }

    /**
     * Returns the token expiry as unix seconds, matching the JWT {@code exp}
     * claim and the {@code showad_expires} cookie payload used by the SDKs.
     */
    public Long getExpirySeconds(String token) {
        TokenClaims claims = decode(token);
        if (claims == null || claims.getExp() == null) {
            return null;
        }
        return claims.getExp();
    }

    /**
     * @deprecated Use {@link #getExpirySeconds(String)}. Kept only to avoid
     *             source breaks for early adopters of the draft SDK.
     */
    @Deprecated
    public Long getExpiryMillis(String token) {
        return getExpirySeconds(token);
    }

    public ValidationResult validateClaims(String token, String expectedCreatorHash, String expectedFingerprint) {
        return validateClaims(token, expectedCreatorHash, expectedFingerprint, new ClaimValidationOptions());
    }

    public ValidationResult validateClaims(
        String token,
        String expectedCreatorHash,
        String expectedFingerprint,
        ClaimValidationOptions options
    ) {
        TokenClaims claims = decode(token);
        if (claims == null) {
            return ValidationResult.invalid("invalid_format");
        }
        if (isExpired(token, options.leewaySeconds)) {
            return ValidationResult.invalid("expired");
        }
        if (expectedCreatorHash == null
            || !fixedTimeEquals(expectedCreatorHash, claims.getCreatorHash())) {
            return ValidationResult.invalid("creator_mismatch");
        }
        if (expectedFingerprint != null
            && !fixedTimeEquals(expectedFingerprint, claims.getFingerprint())) {
            return ValidationResult.invalid("fingerprint_mismatch");
        }
        String iss = claims.getIss();
        if (options.requireIssuer) {
            if (!EXPECTED_ISSUER.equals(iss)) {
                return ValidationResult.invalid("invalid_issuer");
            }
        } else if (iss != null && !EXPECTED_ISSUER.equals(iss)) {
            return ValidationResult.invalid("invalid_issuer");
        }
        return ValidationResult.ok();
    }

    public static Instant nowFromClock(Clock clock) {
        return clock.instant();
    }

    private static boolean fixedTimeEquals(String expected, String actual) {
        if (expected == null || actual == null) {
            return false;
        }
        byte[] a = expected.getBytes(StandardCharsets.UTF_8);
        byte[] b = actual.getBytes(StandardCharsets.UTF_8);
        return MessageDigest.isEqual(a, b);
    }

    private static String padBase64(String value) {
        int rem = value.length() % 4;
        if (rem == 0) {
            return value;
        }
        return value + "====".substring(rem);
    }

    public static final class ClaimValidationOptions {
        public long leewaySeconds = DEFAULT_LEEWAY_SECONDS;
        public boolean requireIssuer = true;

        public ClaimValidationOptions() {}

        public ClaimValidationOptions(long leewaySeconds, boolean requireIssuer) {
            this.leewaySeconds = leewaySeconds;
            this.requireIssuer = requireIssuer;
        }
    }

    public record ValidationResult(boolean valid, String reason) {
        public static ValidationResult ok() { return new ValidationResult(true, null); }
        public static ValidationResult invalid(String reason) { return new ValidationResult(false, reason); }
    }
}

File: wordpress/showad-content-gate/includes/JwtHelper.php
<?php
/**
 * JWT helper — decode and validate JWT tokens.
 *
 * Client-side only: does NOT verify signatures.
 * Signature verification happens on the ShowAd backend via /api/sdk/validate.
 *
 * Defense-in-depth: rejects tokens whose header `alg` is `none` or outside
 * the HS256/HS384/HS512/RS256/RS384/RS512/ES256/ES384 whitelist.
 *
 * @package ShowAd
 */

namespace ShowAd;

if ( ! defined( 'ABSPATH' ) ) {
    exit;
}

class JwtHelper {

    const ISSUER = 'showad-backend';

    const DEFAULT_LEEWAY_SECONDS = 60;

    /**
     * @return string[]
     */
    public static function allowed_algorithms() {
        return array(
            'HS256', 'HS384', 'HS512',
            'RS256', 'RS384', 'RS512',
            'ES256', 'ES384',
        );
    }

    /**
     * Decode a JWT token without verifying the signature.
     *
     * @param string $token JWT token string.
     * @return array|null Decoded claims, or null on failure.
     */
    public static function decode_token( $token ) {
        $parts = explode( '.', $token );
        if ( 3 !== count( $parts ) ) {
            return null;
        }

        $header_json = self::base64url_decode( $parts[0] );
        if ( false === $header_json ) {
            return null;
        }
        $header = json_decode( $header_json, true );
        if ( ! is_array( $header ) || empty( $header['alg'] ) || ! is_string( $header['alg'] ) ) {
            return null;
        }
        if ( ! in_array( $header['alg'], self::allowed_algorithms(), true ) ) {
            return null;
        }

        $payload = self::base64url_decode( $parts[1] );
        if ( false === $payload ) {
            return null;
        }

        $claims = json_decode( $payload, true );
        if ( ! is_array( $claims ) ) {
            return null;
        }

        return $claims;
    }

    /**
     * Check if a token is expired.
     *
     * @param string $token  JWT token.
     * @param int    $leeway Clock-skew tolerance in seconds.
     * @return bool True if expired or invalid.
     */
    public static function is_token_expired( $token, $leeway = self::DEFAULT_LEEWAY_SECONDS ) {
        $claims = self::decode_token( $token );
        if ( null === $claims ) {
            return true;
        }

        $now    = time();
        $leeway = (int) $leeway;

        if ( isset( $claims['exp'] ) && ( (int) $claims['exp'] + $leeway ) < $now ) {
            return true;
        }
        if ( isset( $claims['nbf'] ) && ( (int) $claims['nbf'] - $leeway ) > $now ) {
            return true;
        }
        if ( isset( $claims['iat'] ) && ( (int) $claims['iat'] - $leeway ) > $now ) {
            return true;
        }

        return false;
    }

    /**
     * Get the token expiry as Unix seconds (matches JWT `exp` claim).
     *
     * @param string $token JWT token.
     * @return int|null Expiry in seconds, or null.
     */
    public static function get_token_expiry( $token ) {
        $claims = self::decode_token( $token );
        if ( null === $claims || ! isset( $claims['exp'] ) ) {
            return null;
        }
        return (int) $claims['exp'];
    }

    /**
     * Get seconds until token expiry.
     *
     * @param string $token JWT token.
     * @return int Seconds remaining, or -1 if no expiry.
     */
    public static function get_time_until_expiry( $token ) {
        $claims = self::decode_token( $token );
        if ( null === $claims || ! isset( $claims['exp'] ) ) {
            return -1;
        }
        return max( 0, (int) $claims['exp'] - time() );
    }

    /**
     * Validate token claims against expected values.
     *
     * @param string      $token                JWT token.
     * @param string      $expected_creator_hash Expected creator_hash.
     * @param string|null $expected_fingerprint  Expected fingerprint (optional).
     * @param array       $options { leeway_seconds?: int, require_issuer?: bool }
     * @return array { @type bool $valid, @type string|null $reason }
     */
    public static function validate_token_claims( $token, $expected_creator_hash, $expected_fingerprint = null, $options = array() ) {
        $leeway          = isset( $options['leeway_seconds'] ) ? (int) $options['leeway_seconds'] : self::DEFAULT_LEEWAY_SECONDS;
        $require_issuer  = array_key_exists( 'require_issuer', $options ) ? (bool) $options['require_issuer'] : true;

        $claims = self::decode_token( $token );
        if ( null === $claims ) {
            return array( 'valid' => false, 'reason' => 'invalid_format' );
        }

        if ( self::is_token_expired( $token, $leeway ) ) {
            return array( 'valid' => false, 'reason' => 'expired' );
        }

        $token_creator = isset( $claims['creator_hash'] ) && is_string( $claims['creator_hash'] ) ? $claims['creator_hash'] : '';
        if ( ! empty( $expected_creator_hash ) ) {
            if ( ! hash_equals( (string) $expected_creator_hash, $token_creator ) ) {
                return array( 'valid' => false, 'reason' => 'creator_mismatch' );
            }
        }

        if ( null !== $expected_fingerprint ) {
            $token_fp = isset( $claims['fingerprint'] ) && is_string( $claims['fingerprint'] ) ? $claims['fingerprint'] : '';
            if ( ! hash_equals( (string) $expected_fingerprint, $token_fp ) ) {
                return array( 'valid' => false, 'reason' => 'fingerprint_mismatch' );
            }
        }

        $issuer = isset( $claims['iss'] ) && is_string( $claims['iss'] ) ? $claims['iss'] : '';
        if ( $require_issuer ) {
            if ( self::ISSUER !== $issuer ) {
                return array( 'valid' => false, 'reason' => 'invalid_issuer' );
            }
        } elseif ( ! empty( $issuer ) && self::ISSUER !== $issuer ) {
            return array( 'valid' => false, 'reason' => 'invalid_issuer' );
        }

        return array( 'valid' => true, 'reason' => null );
    }

    /**
     * Extract creator_hash claim from token.
     *
     * @param string $token JWT token.
     * @return string|null
     */
    public static function get_creator_hash_from_token( $token ) {
        $claims = self::decode_token( $token );
        return ( $claims && isset( $claims['creator_hash'] ) ) ? $claims['creator_hash'] : null;
    }

    /**
     * Extract fingerprint claim from token.
     *
     * @param string $token JWT token.
     * @return string|null
     */
    public static function get_fingerprint_from_token( $token ) {
        $claims = self::decode_token( $token );
        return ( $claims && isset( $claims['fingerprint'] ) ) ? $claims['fingerprint'] : null;
    }

    /**
     * Extract session_hash claim from token.
     *
     * @param string $token JWT token.
     * @return string|null
     */
    public static function get_session_hash_from_token( $token ) {
        $claims = self::decode_token( $token );
        return ( $claims && isset( $claims['session_hash'] ) ) ? $claims['session_hash'] : null;
    }

    /**
     * Base64URL decode.
     *
     * @param string $data Base64URL encoded string.
     * @return string|false Decoded data or false.
     */
    private static function base64url_decode( $data ) {
        $remainder = strlen( $data ) % 4;
        if ( $remainder ) {
            $data .= str_repeat( '=', 4 - $remainder );
        }
        $decoded = base64_decode( strtr( $data, '-_', '+/' ), true );
        return $decoded;
    }
}
